File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        R,C=len(grid),len(grid[0])
        visit=set()
        island=0
        def bfs(r,c):
            q=collections.deque()
            visit.add((r,c))
            q.append((r,c))
            while q:
                row,col=q.pop()
                directions=[[1,0],[-1,0],[0,1],[0,-1]]
                for dr,dc in directions:
                    r,c=row+dr,col+dc
                    if (r in range(R) and c in range(C) and grid[r][c]=="1" and (r,c) not in visit):
                        q.append((r,c))
                        visit.add((r,c))
        
        for row in range(R):
            for col in range(C):
                if grid[row][col]=="1" and (row,col) not in visit:
                    bfs(row,col)
                    island+=1
        return island

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
    int dr[4]={0,0,1,-1};
    int dc[4]={1,-1,0,0};
    private:
    void bfs(int sr,int sc,vector<vector<char>>& grid,vector<vector<bool>>&visited,int R,int C){
        queue<pair<int,int>>q;
        visited[sr][sc]=true;
        q.push({sr,sc});
        
        while(!q.empty()){
            auto [r,c]=q.front();
            q.pop();
            for(int d=0;d<4;d++){
                int nr=r+dr[d];
                int nc=c+dc[d];
                if(nr>=0&&nc>=0&&nr<R&&nc<C&&grid[nr][nc]=='1'&&visited[nr][nc]==false){
                    q.push({nr,nc});
                    visited[nr][nc]=true;
                }
            }
        }
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        if(grid.empty()) return 0;
        int R=grid.size(),C=grid[0].size();
        int count=0;
        vector<vector<bool>>visited(R,vector<bool>(C,false));
        for(int row=0;row<R;row++){
            for(int col=0;col<C;col++){
                if(grid[row][col]=='1'&&visited[row][col]==false){
                    bfs(row,col,grid,visited,R,C);
                    count++;
                }
            }
        }
        return count;
    }
};